File: quest-sys/src/debug.cpp
//
// Created by Erich Essmann on 12/03/2025.
//
#include "bindings.h"

namespace quest_sys {
    void setSeeds(rust::Slice<const unsigned> seeds, int numSeeds) {
        ::setSeeds(const_cast<unsigned*>(seeds.data()), numSeeds);
    }
    void setSeedsToDefault() {
        ::setSeedsToDefault();
    }

    std::vector<unsigned> getSeeds() {
        unsigned* seed_ptr;
        ::getSeeds(seed_ptr);
        if (seed_ptr) {
            return {seed_ptr, seed_ptr + ::getNumSeeds()};
        } else {
            return {};
        }
    }

    void invalidQuESTInputError(rust::String msg, rust::String func) {
        ::invalidQuESTInputError(msg.c_str(), func.c_str());
    }

    void setValidationOn() {
        ::setValidationOn();
    }
    void setValidationOff() {
        ::setValidationOff();
    }

    void setValidationEpsilonToDefault() {
        ::setValidationEpsilonToDefault();
    }
    void setValidationEpsilon(qreal eps) {
        ::setValidationEpsilon(eps);
    }
    qreal getValidationEpsilon() {
        return ::getValidationEpsilon();
    }

    void setMaxNumReportedItems(qindex numRows, qindex numCols) {
        ::setMaxNumReportedItems(numRows, numCols);
    }
    void setMaxNumReportedSigFigs(int numSigFigs) {
        ::setMaxNumReportedSigFigs(numSigFigs);
    }

    qindex getGpuCacheSize() {
        return ::getGpuCacheSize();
    }
    void clearGpuCache() {
        ::clearGpuCache();
    }
    rust::String getEnvironmentString() {
        std::array<char, 200> str{};
        ::getEnvironmentString(str.data());
        return {str.data()};
    }
}

File: quest-sys/src/debug.rs
#[cxx::bridge]
pub mod ffi {
    // Types
    unsafe extern "C++" {
        // Common type
        type Quest_Complex = crate::types::ffi::Quest_Complex;
    }

    // Debug
    #[namespace = "quest_sys"]
    unsafe extern "C++" {
        include!("debug.hpp");
        // Random number generators
        fn setSeeds(seeds: &[u32]);
        fn setSeedsToDefault();
        fn getSeeds() -> Vec<u32>;

        // Validation
        fn invalidQuESTInputError(msg: String, func: String);
        fn setValidationOn();
        fn setValidationOff();
        fn setValidationEpsilonToDefault();
        fn setValidationEpsilon(eps: f64);
        fn getValidationEpsilon() -> f64;

        // Reporting
        fn setMaxNumReportedItems(numRows: i64, numCols: i64);
        fn setMaxNumReportedSigFigs(numSigFigs: i32);

        // GPU cache
        fn getGpuCacheSize() -> i64;
        fn clearGpuCache();

        // Environment
        fn getEnvironmentString() -> String;
    }
}
